Фаррахов Фанур 05-204 Вариант 18.

In [1]:
import math
import numpy as np
import sympy as sp
import IPython.display as dp

In [2]:
# TODO:
# Построить графики для оригинальной функции и для ее интерполяции

**Вариант 18.** <br>
$f(x) = 3^{x/5}$ <br>
$a = 3$, $b = 4$ <br>

In [3]:
# Начальные условия
x = sp.symbols('x')
f = 3**(x/5)
df = sp.diff(f)
a = 3
b = 4
X = [15/4, 15/6, 16/(7 * np.pi)] # Контрольные значения x^(i)
Y = [f.subs(x, x_l) for x_l in X]

**Задача 1.**<br>
1) $x_k = a + k \frac{b - a}{4}$, $k = \overline{0,4}$
2) $x_k = \frac{b -a}{2} \cos{\frac{(2k - 1)\pi}{10}} + \frac{b+a}{2}$, $k = \overline{1,5}$

In [4]:
# Первый случай
X1 = [(a + k*(b - a)/4) for k in range(5)]   # x_k
Y1 = [f.subs(x, x_k) for x_k in X1]   # y_k
print(f"x_k = {X1}")
print(f"y_k = {Y1}")

x_k = [3.0, 3.25, 3.5, 3.75, 4.0]
y_k = [1.93318204493176, 2.04234363194535, 2.15766927997459, 2.27950705695478, 2.40822468528069]


In [5]:
# Второой случай
X2 = [((b-a)/2 * np.cos((2*k-1)*np.pi / 10) + (b+a)/2) for k in range(1, 6)]   # x_k
Y2 = [f.subs(x, x_k) for x_k in X2]   # y_k
print(f"x_k = {X2}")
print(f"y_k = {Y2}")

x_k = [3.975528258147577, 3.7938926261462367, 3.5, 3.2061073738537633, 3.024471741852423]
y_k = [2.39531043320922, 2.30159741976771, 2.15766927997459, 2.02274154539847, 1.94360474417031]


Формула вычисления полинома Лагранжа: <br>
$L_n(f, x) = \sum\limits_{k=0}^{n-1}{f(x_k) l_k(x)}$ 
, где 
 $l_k = \prod\limits_{j=0 \atop j \neq k}^{n-1}{\frac{x - x_j}{x_k - x_j}}$ <br>
Формула оценки максимальной погрешности в любой точке $x \in [a, b]$: <br>
$|f(x) - L_n(x)| \leq \frac{\max\limits_{x \in [a, b]}{|f^{n+1}(x)|}}{(n+1)!}$

In [40]:
class Lagrange:    # rename to "LagrangeCalc"

    def __init__(self, X, Y):
        self.n = len(X)
        self.X = np.array(X)
        self.Y = np.array(Y)

    def __l(self, x, k):
        b = [(x - self.X[j]) / (self.X[k] - self.X[j])
             for j in range(self.n) if j != k]
        return np.prod(b, axis=0)

    def interpolate(self, x=x):
        b = [self.Y[k] * self.__l(x, k) for k in range(self.n)]
        self.L = np.sum(b, axis=0)
        self.L = sp.simplify(self.L)
        return self.L

    # |L_5(f, x^i) - f(x^i)| разность в конрольных точках
    def diff_in_control_points(self, X, Y):
        '''
            X - контрольные точки
            Y - значения функции в контрольных точках
        '''
        return np.abs(np.array(Y) - np.array([self.L.subs(x, x_l) for x_l in X]))

    def display(self):
        dp.display(dp.Markdown(rf"$L_{self.n}(f, x) = {sp.latex(self.L)}$"))

    def display_diff_in_conrol_points(self, X, Y):
        dp.display(dp.Markdown(rf"$|L_{self.n}(f, x^i) - f(x^i)| = {list(self.diff_in_control_points(X, Y))}$, $i = 1, 2, 3$"))

**Вычисление на первых узлах**

In [41]:
lp1 = Lagrange(X1, Y1)
lp1.interpolate()
lp1.display()

$L_5(f, x) = 0.000209647567984916 x^{4} + 0.000882482474935387 x^{3} + 0.0274087663992759 x^{2} + 0.213841099584897 x + 1.00417136875876$

In [42]:
lp1.display_diff_in_conrol_points(X, Y)

$|L_5(f, x^i) - f(x^i)| = [6.16662276797797e-12, 6.24694287809646e-6, 0.00131277136856656]$, $i = 1, 2, 3$

In [43]:
# TODO: Дать оценку приближения

**Вычисление на вторых узлах**

In [44]:
lp2 = Lagrange(X2, Y2)
lp2.interpolate()
lp2.display()

$L_5(f, x) = 0.000209647566819626 x^{4} + 0.000882482449696909 x^{3} + 0.0274087654315736 x^{2} + 0.213841071503339 x + 1.00417148015913$

In [45]:
lp2.display_diff_in_conrol_points(X, Y)

$|L_5(f, x^i) - f(x^i)| = [9.06899977337616e-9, 6.28165133997882e-6, 0.00131286181546142]$, $i = 1, 2, 3$

In [12]:
# TODO: Дать оценку приближения

***
**Задача 2.** <br>
$x_k = a + k \frac{b-a}{4}$, $k = \overline{0, 4}$, как в задачи 1 пункт 1

Формула вычисления полинома Лагранжа в форме Ньютона:
$L_4(f, x) = f(x_1) + (x - x_1) f(x_1, x_2) + (x - x_1)(x - x_2) f(x_1, x_2, x_3) + (x - x_1)(x - x_2)(x - x_3) f(x_1, x_2, x_3, x_4) + (x - x_1)(x - x_2)(x - x_3)(x - x_4) f(x_1, x_2, x_3, x_4, x_5)$ <br>
,где <br>
$f(x_k, x_{k+1}) = \frac{f(x_{k+1}) - f(x_k)}{x_{k+1} - x_k}$ <br>
$f(x_{k-1}, x_k, x_{k+1}) = \frac{f(x_k, x_{k+1}) - f(x_{k-1}, x_k)}{x_{k+1} - x_{k-1}}$ <br> <br>
$
f(x_0, x_1, \ldots, x_n)=\sum_{j=0}^n\dfrac{f(x_j)}{\prod\limits_{i=0\atop i\neq j}^n(x_j-x_i)} \qquad (*)
$ <br>
Доказательство ф-лы $(*)$: <br>
...

In [46]:
class Newton:

    def __init__(self, X, Y):
        self.n = len(X)
        self.X = np.array(X)
        self.Y = np.array(Y)

    def _f(self, n):
        '''
            n - число аргументов в функции f. Например, нужно найти f(x1, x2, x3), то n будет равно 3, а если же нужно найти f(x1, x2, x3, x4), то n = 4
        '''
        b = [self.Y[j] / self._w(n, j) for j in range(n)]
        return np.sum(b, axis=0)


    def _w(self, n, j):
        b = [(self.X[j] - self.X[i]) for i in range(n) if i != j]
        return np.prod(b, axis=0)
        

    def interpolate(self, x=x):
        omega = [(x - self.X[i]) for i in range(self.n)]
        b = [self._f(n) * np.prod(omega[:n-1], axis=0) for n in range(1, self.n + 1)]
        self.L = np.sum(b, axis=0)
        self.L = sp.simplify(self.L)
        return self.L

    def diff_in_control_points(self, X, Y):
        '''
            X - контрольные точки
            Y - значения функции в контрольных точках
        '''
        return np.abs(np.array(Y) - np.array([self.L.subs(x, x_l) for x_l in X]))

    def display(self):
        dp.display(dp.Markdown(rf"$L_{self.n}(f, x) = {sp.latex(self.L)}$"))

    def display_diff(self, X, Y):
        dp.display(dp.Markdown(rf"$|L_{self.n}(f, x^i) - f(x^i)| = {list(self.diff_in_control_points(X, Y))}$, $i = 1, 2, 3$"))
    

In [15]:
print(f"x_k = {X1}")
print(f"y_k = {Y1}")

x_k = [3.0, 3.25, 3.5, 3.75, 4.0]
y_k = [1.93318204493176, 2.04234363194535, 2.15766927997459, 2.27950705695478, 2.40822468528069]


In [16]:
nlp = Newton(X1, Y1)

In [17]:
nlp.interpolate()
nlp.display()

$L_5(f, x) = 0.000209647567981364 x^{4} + 0.000882482475342172 x^{3} + 0.0274087663968203 x^{2} + 0.213841099586788 x + 1.00417136875929$

In [18]:
X3 = [-1, 0, 1, 3]
Y3 = [0, -1, 0, 8]

In [19]:
nlp2 = Newton(X3, Y3)

In [20]:
nlp2.interpolate()
nlp2.display()

$L_4(f, x) = 1.0 x^{2} - 1.0$

**Задача 3.** <br>
$x_k = a + k \frac{b - a}{4}$, $k = \overline{0, 4}$

In [21]:
dY1 = [df.subs(x, x_k) for x_k in X1]

In [22]:
print(f"x_k  = {X1}")
print(f"y_k  = {Y1}")
print(f"y'_k = {dY1}")

x_k  = [3.0, 3.25, 3.5, 3.75, 4.0]
y_k  = [1.93318204493176, 2.04234363194535, 2.15766927997459, 2.27950705695478, 2.40822468528069]
y'_k = [0.386636408986353*log(3), 0.40846872638907*log(3), 0.431533855994919*log(3), 0.455901411390956*log(3), 0.481644937056138*log(3)]


Формула для вычисления полинома Эрмита-Фейера: <br>
$H_{m - 1}(f, x) = L_n(f, x) + (x - x_1)\ldots(x - x_n) H_{m-n-1}(x) = L_n(f, x) + \omega_n(x) H_{m-n-1}(x)$, где $L_n(f, x)$ - обычный полином Лагранжа, и $H_{m-n-1}(x)$ находится следующим способом.

Для начала найдем производную $H_{m - 1}(f, x)$. 
$$H_{m - 1}'(f, x) = L_n'(f, x) + \omega_n'(x) H_{m-n-1}(x) + \omega_n(x) H_{m-n-1}'(x)$$
Теперь подставим узловые точки $x_k$ в полином $H_{m - 1}'(f, x)$.
$$H_{m - 1}'(f, x_k) = L_n'(f, x_k) + \omega_n'(x_k) H_{m-n-1}(x_k) + \omega_n(x_k) H_{m-n-1}'(x_k), \text{где } \omega_n(x_k)=0 \quad k = \overline{1,n}$$
$$\Downarrow$$
$$H_{m - 1}'(f, x_k) = L_n'(f, x_k) + \omega_n'(x_k) H_{m-n-1}(x_k) = f'(x_k)$$
$$\Downarrow$$
$$H_{m-n-1}(x_k) = \frac{f'(x_k) - L_n'(f, x_k)}{\omega_n'(x_k)} = g(x_k)$$
$H_{m-n-1}(x) = L_n^*(f, x) + \omega_n(x) H_{m-2n-1}(x)$, $\quad H_{m-2n-1}(x)$ находится из оставшихся условий <br>
В нашем случае, т.е. крастность каждого узла = 2, $H_{m-2n-1}(x) = 0$, потому что нет оставшихся условий. 
Поэтому $H_{m-n-1}(x) = L_n^*(f, x)$.<br>
$L_n^*(f, x)$ - какой-то другой полином Лагранжа, построенный на точках $\quad {(x_1, g(x_1)), (x_2, g(x_2)), \ldots, (x_n, g(x_n))}$

In [29]:
# TODO:
# Как-то покрасивей переписать код

In [79]:
class Hermit:

    def __init__(self, X, Y, dY):
        self.n = len(X)
        self.X = X
        self.Y = Y
        self.dY = dY

    def _H(self, x):
        dL = sp.diff(self._L)
        Y2 = [(self.dY[i] - sp.diff(self._L).subs(x, self.X[i])) / sp.diff(self._w).subs(x, self.X[i]) for i in range(self.n)]
        # print(f"Y2 = {Y2}")
        return Lagrange(self.X, Y2).interpolate()
        
    def interpolate(self, x=x):
        self._L = Lagrange(self.X, self.Y).interpolate()
        self._w = np.prod([(x - self.X[k]) for k in range(self.n)])
        self.H = self._L + self._w * self._H(x)
        self.H = sp.expand(self.H)
        return self.H
        
    def display(self):
        dp.display(dp.Markdown(rf"$H_{2 * self.n - 1}(f, x) = {sp.latex(sp.simplify(self.H))}$"))

In [80]:
her = Hermit(X1, Y1, dY1)

In [81]:
hermit = her.interpolate()
her.display()

$H_9(f, x) = - 7.31142790755257 \cdot 10^{-9} x^{9} + 2.31486865232 \cdot 10^{-7} x^{8} - 3.24234252957467 \cdot 10^{-6} x^{7} + 2.67065118180199 \cdot 10^{-5} x^{6} - 0.000135092489841782 x^{5} + 0.000584107078211105 x^{4} + 0.000634997332326748 x^{3} + 0.0258310505951158 x^{2} + 0.218250390613321 x + 1.00056837297416$

***
Проверка моей догадки, что <br>
$H_{m-n-1}(x) = L_n(f,x)$
Для задачи: <br>
|   |x_1|x_2|
|---|---|---|
|y  | -1|  1|
|y' |1/2|-1|

In [34]:
X4  = [0, 1]
Y4  = [-1, 1]
dY4 = [1/2, -1]

In [35]:
hp = Hermit(X4, Y4, dY4)

In [36]:
hp.interpolate()

-4.5*x**3 + 6.0*x**2 + 0.5*x - 1

Вычисленно в maple: <br>
$0.9863668486+ 0.000000177225121185227\,{x}^{9}- 0.000005600378880\,{x
}^{8}+ 0.2550927871\,x- 0.01657986874\,{x}^{2}+ 0.02906881170\,{x}^{3} 
- 0.01165162609\,{x}^{4}+ 0.003369756621\,{x}^{5}- 0.0006415856096\,{x
}^{6}+ 0.00007855522009\,{x}^{7}
$ <br>
Вычисленно мной: <br>
$- 7.31142790755257 \cdot 10^{-9} x^{9} + 2.31486865232 \cdot 10^{-7} x^{8} - 3.24234252957467 \cdot 10^{-6} x^{7} + 2.67065118180199 \cdot 10^{-5} x^{6} - 0.000135092489841782 x^{5} + 0.000584107078211105 x^{4} + 0.000634997332326748 x^{3} + 0.0258310505951158 x^{2} + 0.218250390613321 x + 1.00056837297416$